cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -846,71 +700,76 @@ void process_and_fill_codes(const handle_t& handle,
                             std::variant<IdxT, const IdxT*> src_offset_or_indices,
                             const uint32_t* new_labels,
                             IdxT n_rows,
-                            rmm::mr::device_memory_resource* device_memory,
-                            rmm::mr::device_memory_resource* managed_memory)
+                            rmm::mr::device_memory_resource* mr)
 {
   pq_int_vec_exts pq_extents = make_extents<size_t>(index.pq_dataset().extent(0),
                                                     index.pq_dataset().extent(1),
                                                     index.pq_dataset().static_extent(2));
   auto pq_dataset            = make_mdspan<pq_vec_t, size_t, row_major, false, true>(
     reinterpret_cast<pq_vec_t*>(index.pq_dataset().data_handle()), pq_extents);
 
-  rmm::device_uvector<uint32_t> new_cluster_sizes_buf(
-    index.n_lists(), handle.get_stream(), managed_memory);  // host access in compute_pq_codes
-  auto new_cluster_sizes = new_cluster_sizes_buf.data();
-  rmm::device_uvector<IdxT> new_cluster_offsets(
-    index.n_lists() + 1, handle.get_stream(), managed_memory);  // host access in compute_pq_codes
-  raft::stats::histogram<uint32_t, IdxT>(raft::stats::HistTypeAuto,
-                                         reinterpret_cast<int32_t*>(new_cluster_sizes),
-                                         IdxT(index.n_lists()),
+  auto stream = handle.get_stream();
+  auto new_vectors_residual =
+    make_device_mdarray<float>(handle, mr, make_extents<IdxT>(n_rows, index.rot_dim()));
+
+  {
```

Comment:
Why can't we use `select_residuals` instead of these blocks? Consider unifying these in a follow up work.

Reply:
That's because we have different access patterns. In `select_residuals` we select rows from the dataset, in the new code go through all rows, but select the centers (cluster ids). It make sense to put the new code in the similar function though!